a2s/byteio.py

```diff
@@ -1,80 +1,95 @@
-import struct
+from __future__ import annotations
+
 import io
+import struct
+from typing import TYPE_CHECKING, Any, Optional, Tuple, Union
 
 from a2s.exceptions import BufferExhaustedError
 
+if TYPE_CHECKING:
+    from typing_extensions import Literal
+
+STRUCT_OPTIONS = Literal[
+    "x", "c", "b", "B", "?", "h", "H", "i", "I", "l", "L", "q", "Q", "n", "N", "e", "f", "d", "s", "p", "P"
+]
 
 
-class ByteReader():
-    def __init__(self, stream, endian="=", encoding=None):
-        self.stream = stream
-        self.endian = endian
-        self.encoding = encoding
+class ByteReader:
+    __slots__ = (
+        "stream",
+        "endian",
+        "encoding",
+    )
 
-    def read(self, size=-1):
+    def __init__(self, stream: io.BytesIO, endian: str = "=", encoding: Optional[str] = None) -> None:
+        self.stream: io.BytesIO = stream
+        self.endian: str = endian
+        self.encoding: Optional[str] = encoding
+
+    def read(self, size: int = -1) -> bytes:
         data = self.stream.read(size)
         if size > -1 and len(data) != size:
             raise BufferExhaustedError()
 
         return data
 
-    def peek(self, size=-1):
+    def peek(self, size: int = -1) -> bytes:
         cur_pos = self.stream.tell()
         data = self.stream.read(size)
         self.stream.seek(cur_pos, io.SEEK_SET)
         return data
 
-    def unpack(self, fmt):
-        fmt = self.endian + fmt
+    def unpack(self, fmt: STRUCT_OPTIONS) -> Tuple[Any, ...]:
```

Comment:
This is going too far, it defines possible format arguments much stricter than they need to be in an inherently unsafe module. Simply typing this as `str` is better.

Reply:
Fixed locally and will be pushed in a moment!